proposals/break-continue-in-inline-lambdas.md

```diff
@@ -60,57 +64,45 @@ So the Motivation is the following:
+        if (i == 10) continue // Where to continue?
     }
 }
 ```
 
-The alternative approach may be to make `break`/`continue` apply to the closest `for`, `while` or "loop-like" functions. We will
-never allow unlabeled `break`/`continue` for "loop-like" functions. The rationale is simple: if Kotlin allowed unlabeled
-`break`/`continue` for "loop-like" functions then it would mean that for such code
+However, it seems that both features can be implemented without introducing ambiguities.
 
+First of all, we should never allow unlabeled `break`/`continue` for "loop-like" functions because users
+wouldn't be able to understand whether it is applied to a certain function without checking its implementation.
 ```kotlin
-while (cond) {
-    someInlineFunctionThatAcceptsLambda {
-        break
+foo { bar { break } } // which function is loop-like?
+```
+That leave us only with the labeled syntax option for "loop-like" functions:
+```kotlin
+for (i in 0 until 100) {
+    (0 until 200).forEach {
+        if (it == 42) break@forEach // Obviously, it breaks `forEach`
+        if (i == 10) continue // No label, so it applies to the 'for' loop
     }
 }
 ```
 
-users wouldn't be able to understand whether `break` breaks from `someInlineFunctionThatAcceptsLambda` or from `while` without
-looking into `someInlineFunctionThatAcceptsLambda` implementation. It complicates reasoning about `break` and `continue`
-statements. In other words, it's hard to distinguish "loop-like" functions from other lambdas if the call-site doesn't differ
-syntactically.
-
-Also, right now we have such an invariant in Kotlin: a pair of **keywords** `break`/`continue` tied to the closest `while` or
-`for` **keywords**. `return` **keyword** is tied to the closest `fun` **keyword**. The proposed alternative solution breaks this
-invariant.
-
-* **Q**: Why not make labels mandatory for non-local `break`/`continue`? That would eliminate `forEach` puzzler.
-* **A**: Because Kotlin doesn't require labels for non-local `return`. For sake of consistency, We should either require labels
-  for both `break`/`continue` and `return` cases or don't require the labels again for the both cases.
-
-* **Q**: Ok, why not make labels mandatory for both non-local `return` and non-local `break`/`continue` then?
-* **A**: We can't make non-local `return` to be obligatory labeled because it would break inline functions first-class
-  citizenship in language constructs. E.g. we value "return if not null" idiomatic Kotlin construct `nullable?.let { return it }`.
-  If non-local `return` were obligatory labeled then the code would be much more verbose
-  `nullable?.let { return@functionName it }`.
-
-It worth noticing that we don't abandon the idea to provide some "break"/"continue" mechanism for "loop-like" functions
-[KT-19748](https://youtrack.jetbrains.com/issue/KT-19748). It just that we agree that we will never support unlabeled non-local
-`break`/`continue` for "loop-like" functions.
```

Comment:
@strangepleasures I think that you dropped an important part from the KEEP.

1. `break`/`continue` <-> `for`/`while` invariant notice
2. The reasoning that answers the question: "why are we ok with introducing a new language feature that brings `forEach` vs `for`/`while` puzzler" (Well, I think that that we can move this reasoning into a "Tooling for puzzling code" but I don't think that we should drop this reasoning)